src/main/java/dk/kb/util/yaml/YAML.java

```diff
@@ -668,52 +684,293 @@ public boolean containsKey(Object path) throws NullPointerException {
+     */
+    private void traverse(List<String> yPath, Object yaml, MultipleValuesVisitor visitor) {
+        if (yaml instanceof Map) {
+            traverseMap(yPath, yaml, visitor);
+        } else if (yaml instanceof List) {
+            traverseList(yPath, yaml, visitor);
+        }
+    }
+
+    /**
+     * Traverse a YAML List recursively and collect matching paths to the input visitor.
+     * The implementation supports retrieval of multiple matching results.
+     * <p> <br>
+     * It supports placeholders such as foo[] and foo[*], which both of them returns all elements in this list.
+     * This can be used in conjunction with further traversal such as foo[].bar. The implementation also supports the
+     * syntax foo[last], which will return the last element in the list. Standard index based lookup such as foo[2] is
+     * also supported. The implementation supports dot-array syntax as well. This means that the queries foo[bar] and
+     * foo.[bar] returns the same value.
+     * <p> <br>
+     * The implementation supports conditional lookups as well. This means that a query for zoo.[foo=bar].baz returns
+     * all values for baz, that are a child of zoo and has the sibling-key foo with the value bar.
+     * @param yPath a list of path elements. This list contains all parts of a specified path, which is most likely
+     *              delivered through the {@link #visit(Object path, YAML yaml)}-method.
+     * @param yaml the current place in the YAML file being traversed. This should be an instance of a List.
+     * @param visitor used to collect values that match the given path. This visitor only collects values to its
+     *                internal list of extracted values: {@link MultipleValuesVisitor#extractedValues}.
+     */
+    private void traverseList(List<String> yPath, Object yaml, MultipleValuesVisitor visitor) {
+        List<Object> list = (List<Object>) yaml;
+
+        Matcher matcher = ARRAY_ELEMENT.matcher(yPath.get(0));
+        final String arrayElementIndex;
+        if (matcher.matches()) { // foo.bar[2]
+            arrayElementIndex = matcher.group(2);
+        } else {
+            arrayElementIndex = null;
+        }
+
+        if (arrayElementIndex != null){
+            switch (arrayElementIndex) {
+                case "*":
+                case "":
+                    // Set yPath entry to * as [] and [*] are to be treated equal.
+                    yPath.set(0, "*");
+                    convertListToMapAndTraverse(yPath, visitor, list);
+                    break;
+                case "last":
+                    // Set yPath entry to the last index of the list.
+                    yPath.set(0, String.valueOf(list.size()-1));
+                    convertListToMapAndTraverse(yPath, visitor, list);
+                    break;
+                default:
+                    // Set yPath entry as index number
+                    yPath.set(0, arrayElementIndex);
```

Comment:
Fixed here, but still present in `removeBracketsFromPathElement` (line 929).

The `yPath` is used a lot throughout the code. It might be worth considering adding a helper class to encapsulate that, with methods like `getFirst` for probing and `removeFirst` that return a new instance without the first element. That would also get rid of the helper-methods for `yPath` minipulation in YAML.